sklearn/cluster/_hdbscan/tests/test_hdbscan.py

```diff
@@ -0,0 +1,381 @@
+
+    data = generate_noisy_data()
+
+    labels_prims = hdbscan(data, algorithm="generic")[0]
+    labels_boruvka = hdbscan(data, algorithm=f"boruvka_{tree}")[0]
+
+    num_mismatches = homogeneity(labels_prims, labels_boruvka)
+
+    assert (num_mismatches / float(data.shape[0])) < 0.15
+
+    labels_prims = HDBSCAN(algorithm="generic").fit_predict(data)
+    labels_boruvka = HDBSCAN(algorithm=f"boruvka_{tree}").fit_predict(data)
+
+    num_mismatches = homogeneity(labels_prims, labels_boruvka)
+
+    assert (num_mismatches / float(data.shape[0])) < 0.15
+
+
+def test_hdbscan_badargs():
+    assert_raises(ValueError, hdbscan, X="fail")
+    assert_raises(ValueError, hdbscan, X=None)
+    assert_raises(ValueError, hdbscan, X, min_cluster_size="fail")
+    assert_raises(ValueError, hdbscan, X, min_samples="fail")
+    assert_raises(ValueError, hdbscan, X, min_samples=-1)
+    assert_raises(ValueError, hdbscan, X, metric="imperial")
+    assert_raises(ValueError, hdbscan, X, metric=None)
+    assert_raises(
+        ValueError, hdbscan, X, metric="precomputed", algorithm="boruvka_kdtree"
+    )
+    assert_raises(
+        ValueError, hdbscan, X, metric="precomputed", algorithm="prims_kdtree"
+    )
+    assert_raises(
+        ValueError, hdbscan, X, metric="precomputed", algorithm="prims_balltree"
+    )
+    assert_raises(
+        ValueError, hdbscan, X, metric="precomputed", algorithm="boruvka_balltree"
+    )
+    assert_raises(ValueError, hdbscan, X, alpha=-1)
+    assert_raises(ValueError, hdbscan, X, alpha="fail")
+    assert_raises(Exception, hdbscan, X, algorithm="something_else")
+    assert_raises(TypeError, hdbscan, X, metric="minkowski", p=None)
+    assert_raises(ValueError, hdbscan, X, leaf_size=0)
+
+
+def test_hdbscan_sparse():
+
+    sparse_X = sparse.csr_matrix(X)
+
+    labels = HDBSCAN().fit(sparse_X).labels_
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == 3
+
+
+def test_hdbscan_caching():
+
+    cachedir = mkdtemp()
+    labels1 = HDBSCAN(memory=cachedir, min_samples=5).fit(X).labels_
+    labels2 = HDBSCAN(memory=cachedir, min_samples=5, min_cluster_size=6).fit(X).labels_
+    n_clusters1 = len(set(labels1)) - int(-1 in labels1)
+    n_clusters2 = len(set(labels2)) - int(-1 in labels2)
+    assert n_clusters1 == n_clusters2
+
+
+def test_hdbscan_centroids_medoids():
+    centers = [(0.0, 0.0), (3.0, 3.0)]
+    H, _ = make_blobs(n_samples=1000, random_state=0, centers=centers, cluster_std=0.5)
+    clusterer = HDBSCAN().fit(H)
+
+    for idx, center in enumerate(centers):
+        centroid = clusterer.weighted_cluster_centroid(idx)
+        assert_array_almost_equal(centroid, center, decimal=1)
+
+        medoid = clusterer.weighted_cluster_medoid(idx)
+        assert_array_almost_equal(medoid, center, decimal=1)
+
+
+def test_hdbscan_no_centroid_medoid_for_noise():
+    clusterer = HDBSCAN().fit(X)
+    assert_raises(ValueError, clusterer.weighted_cluster_centroid, -1)
```

Comment:
To be consistent with the rest of scikit-learn, let's use `with pytest.raises`.

This comment applies to all the other places with `assert_raises`.